### PR TITLE
Updated reference to gemmforge (to version 0.0.203)

### PR DESCRIPTION
**Pull request by ravil-mobile**
* gemmforge changed its interface after refactoring
* improved yateto caching of kernel generated by gemmforge
* forced useris **always** initialize streams for batch computing
(no default option anymore). `nullptr` is used by CUDA and HIP as a
default stream. Therefore, a value `0xffffffff` is used instead.
This is necessary because SYCL model uses objects of Queue class
to launch kernels. Queues have a similar meaning as streams
in CUDA/HIP but also has some information about a selected device,
memory, etc.

[Link](http://vmbungartz10.informatik.tu-muenchen.de/seissol/view/Yateto/job/yateto-codegen/10/console) to the test results

### CODE REVIEW

File: yateto/codegen/common.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ class BatchedOperationsAux:
   NUM_ELEMENTS_NAME = 'numElements'
   EXTRA_OFFSET_NAME = 'extraOffset'
   STREAM_PTR_NAME = 'streamPtr'
+  FORBIDDEN_STREAM = 'reinterpret_cast<void*>(0xffffffff)'
```
Comment (uphoffc) [inline yateto/codegen/common.py:94]: Is that a good idea?

**Comment by ravil-mobile** [comment]
Here is a current solution in `Device`
```
void *ConcreteAPI::getDefaultStream() { return nullptr; }
``` 

When I create a dedicated stream to replace nullptr (`cudaStreamCreate(&m_defaultStream);`)
```
void *ConcreteAPI::getDefaultStream() { return m_defaultStream; }
```
I get incorrect numerical results. But I think it is possible to get rid off `0xffffffffffffffff` in the future

**Comment by ravil-mobile** [comment]
Btw, maybe?
```
reinterpret_cast<void*>(std::numeric_limits<size_t>::max());
```

**Comment by uphoffc** [comment]
But assert(streamPtr != nullptr); would yield the same behaviour, right? I don't quite get the difference between release and debug, because all kernels in yateto are unsafe in release mode, i.e. programming mistakes are not caught.

**Comment by ravil-mobile** [comment]
Yes, you're right that programming mistakes are not caught in Release mode. There is a difference between, for example, `assert(TinvT != nullptr);` and `assert(streamPtr != nullptr)`. In the first case (in Release mode and assuming that both `TinvT` and `streamPtr` are equal to `nullptr`) a program will crash because `nullptr` is illegal memory address. But in the second case, CUDA/HIP will treat `nullptr` as a default stream. I assume that CUDA/HIP run-time has something like

```
struct CUstream_st {
...
} default;
typedef struct CUstream_st *cudaStream_t;

void cudaStreamCreate(cudaStream_t& ptr) {
  ptr = new CUstream_st;
  // change some fields
}
...

void details::cuda(...,cudaStream_t userStream) {
  cudaStream_t stream = &default;
  if (userStream != nullptr) {
    stream = userStream;
  }
}
```

In short, in the second case, a program won't crash but kernels scheduling will be changed which may lead to different numerical results and it will be difficult to debug. I mean it will take time to grasp why numerics is correct in debug but not in realease _in the future_ during a debugging process. I would rather prefer a program to crash immediately in both debug and release modes caused by  `reinterpret_cast<void*>(std::numeric_limits<size_t>::max())`.

What do you think?

**Comment by uphoffc** [comment]
Ok, but then I'd use std::numeric_limits<uintptr_t>::max()

**Review by uphoffc — APPROVED** [review]
